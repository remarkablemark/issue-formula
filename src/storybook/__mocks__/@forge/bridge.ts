import { fn } from '@storybook/test';

export const requestJira = fn();

export const view = {
  submit: fn(),
  close: fn(),
  refresh: fn(),
  createHistory: fn(),
  getContext: fn(),
  theme: {
    enable: fn(),
  },
};
