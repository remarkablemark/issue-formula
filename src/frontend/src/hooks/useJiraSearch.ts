import { requestJira } from '@forge/bridge';
import { useEffect, useState } from 'react';

import { log } from '../helpers';
import { FormValues, Issue } from '../types';

export function useJiraSearch(formValues: FormValues) {
  const [hasLoaded, setHasLoaded] = useState(false);
  const [issues, setIssues] = useState<Issue[]>([]);

  useEffect(() => {
    if (hasLoaded || !formValues.jql.length) {
      return;
    }

    const requests = formValues.jql.map((jql, index) =>
      requestJira('/rest/api/3/search', {
        method: 'POST',
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          jql,
          maxResults:
            formValues.function[index].value === 'COUNT' ? 0 : undefined,
        }),
      })
        .then((response) => response.json())
        .catch(log.error),
    );

    Promise.all(requests)
      .then((issues) => {
        setIssues(issues);
      })
      .catch(log.error)
      .finally(() => setHasLoaded(true));
  }, [formValues, hasLoaded, issues]);

  return {
    isLoading: !hasLoaded,
    issues,
  };
}
