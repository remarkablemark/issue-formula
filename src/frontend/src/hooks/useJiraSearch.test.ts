import { requestJira } from '@forge/bridge';
import { renderHook, waitFor } from '@testing-library/react';

import type { FormValues } from '../types';
import { useJiraSearch } from './useJiraSearch';

const mockedRequestJira = jest.mocked(requestJira);

it('returns when formValues.jql is empty', () => {
  const { result } = renderHook(() => useJiraSearch({} as FormValues));
  expect(result.current).toEqual({
    isLoading: false,
    issues: [],
  });
});

it.each(['COUNT', 'SUM'])(
  'requests jira when function value is %p',
  async (value) => {
    mockedRequestJira.mockImplementationOnce(async () => {
      return { json: jest.fn() } as unknown as Response;
    });

    const { result } = renderHook(() =>
      useJiraSearch({
        function: [{ value }],
        jql: ['test'],
      } as FormValues),
    );

    await waitFor(() => {
      expect(result.current).toEqual({
        isLoading: false,
        issues: [],
      });
    });

    expect(mockedRequestJira).toHaveBeenCalledTimes(1);
    expect(mockedRequestJira.mock.calls[0]).toMatchSnapshot();
  },
);
