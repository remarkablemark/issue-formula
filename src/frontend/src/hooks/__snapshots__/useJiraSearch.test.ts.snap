// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requests jira when function value is "COUNT" 1`] = `
[
  "/rest/api/3/search",
  {
    "body": "{"jql":"test","maxResults":0}",
    "headers": {
      "Accept": "application/json",
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
]
`;

exports[`requests jira when function value is "SUM" 1`] = `
[
  "/rest/api/3/search",
  {
    "body": "{"jql":"test"}",
    "headers": {
      "Accept": "application/json",
      "Content-Type": "application/json",
    },
    "method": "POST",
  },
]
`;
